README.md
# ChatGPTForVSCode README

> ChatGPT extension for VSCode 测试版本

## Features

![pic](https://s2.loli.net/2023/08/23/JAcPbyG3n1HvVFU.gif)
command + shift + P: 调出命令面板

* 输入命令：”GPT:绑定APIKey“
* 输入命令：”GPT:询问 GPT“
* 选择代码后：”GPT:说明这段代码存在的问题“
* 选择代码后：”GPT:优化这段代码“
* 选择代码后：”GPT:解释这段代码“
* 选择代码后：”GPT:重构这段代码“

## Setup

Make sure to install the dependencies:

```bash
# yarn
yarn install

yarn test 

yarn vscode:package
```

## Documents

如果想参考使用这个工程，可以查看这个文档：[使用文档](https://chenxming.github.io/2023/08/23/从零开发一款ChatGPT%20VSCode插件/)

## Support:

如果对你有帮助，可以支持一下(＾－＾)V：
<img src="https://s2.loli.net/2023/08/23/WOByR54qvuSFsY3.jpg" alt="pic" style="zoom: 33%;" />

CHANGELOG.md
# Change Log

All notable changes to the "ChatGPTForVSCode" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

out/extension.js
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = void 0;
const vscode = require("vscode");
const view_provider = require("./view_provider");
function activate(context) {
    const chatViewProvider = new view_provider.default(context);
    context.subscriptions.push(vscode.commands.registerCommand('chatgpt-for-vscode.askGPT', askChatGPT), vscode.commands.registerCommand('chatgpt-for-vscode.whyBroken', askGPTWhyBroken), vscode.commands.registerCommand('chatgpt-for-vscode.explainCode', askGPTToExplain), vscode.commands.registerCommand('chatgpt-for-vscode.refactor', askGPTToRefactor), vscode.commands.registerCommand('chatgpt-for-vscode.optimizeCode', askGPTToOptimize), vscode.commands.registerCommand('chatgpt-for-vscode.setAPIKey', resetToken), vscode.window.registerWebviewViewProvider("chatgpt-for-vscode.view", chatViewProvider, {
        webviewOptions: { retainContextWhenHidden: true }
    }));
    async function askGPTWhyBroken() { await askChatGPT('说明下面的代码会出现什么问题?'); }
    async function askGPTToExplain() { await askChatGPT('请帮我解释一下下面的代码?'); }
    async function askGPTToRefactor() { await askChatGPT('帮我重构下面的代码'); }
    async function askGPTToOptimize() { await askChatGPT('帮我优化下面的代码'); }
    async function resetToken() {
        await chatViewProvider.ensureApiKey();
    }
    async function askChatGPT(userInput) {
        let editor = vscode.window.activeTextEditor;
        if (editor) {
            const selectedCode = editor.document.getText(vscode.window.activeTextEditor?.selection);
            if (selectedCode.length) {
                chatViewProvider.sendOpenAiApiRequest(userInput, selectedCode);
                vscode.window.showInformationMessage(selectedCode);
            }
            else {
                vscode.window.showInformationMessage(`请选中一段代码`);
            }
        }
    }
}
exports.activate = activate;
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activate = void 0;
import * as vscode from 'vscode';
const view_provider = require("./view_provider");

export function activate(context: vscode.ExtensionContext) {

	const chatViewProvider = new view_provider.default(context);

	context.subscriptions.push(
	vscode.commands.registerCommand('chatgpt-for-vscode.askGPT', askChatGPT), 
	vscode.commands.registerCommand('chatgpt-for-vscode.whyBroken', askGPTWhyBroken), 
	vscode.commands.registerCommand('chatgpt-for-vscode.explainCode', askGPTToExplain), 
	vscode.commands.registerCommand('chatgpt-for-vscode.refactor', askGPTToRefactor), 
	vscode.commands.registerCommand('chatgpt-for-vscode.optimizeCode', askGPTToOptimize), 
	 vscode.commands.registerCommand('chatgpt-for-vscode.setAPIKey', resetToken), 
	vscode.window.registerWebviewViewProvider("chatgpt-for-vscode.view", chatViewProvider, {
		webviewOptions: { retainContextWhenHidden: true }})
	);
	async function askGPTWhyBroken() { await askChatGPT('说明下面的代码会出现什么问题?'); }
	async function askGPTToExplain() { await askChatGPT('请帮我解释一下下面的代码?'); }
	async function askGPTToRefactor() { await askChatGPT('帮我重构下面的代码'); }
	async function askGPTToOptimize() { await askChatGPT('帮我优化下面的代码'); }
	async function resetToken() {
		await chatViewProvider.ensureApiKey();
	}

	async function askChatGPT(userInput: string) {
      
        let editor = vscode.window.activeTextEditor;
        if (editor) {
            const selectedCode = editor.document.getText(vscode.window.activeTextEditor?.selection);
			if(selectedCode.length) {
			    chatViewProvider.sendOpenAiApiRequest(userInput, selectedCode);
				vscode.window.showInformationMessage(selectedCode);
			}else {
				vscode.window.showInformationMessage(`请选中一段代码`);
			}
        }
    }
}

export function deactivate() {}


out/view_provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vscode = require("vscode");
const undici_1 = require("undici");
class ChatGptViewProvider {
    constructor(context) {
        this.context = context;
    }
    resolveWebviewView(webviewView, _context, _token) {
        this.webView = webviewView;
        webviewView.webview.options = {
            enableScripts: true,
            localResourceRoots: [this.context.extensionUri]
        };
        webviewView.webview.html = this.getHtml(webviewView.webview);
        webviewView.webview.onDidReceiveMessage(data => {
            if (data.type === 'askChatGPT') {
                this.sendOpenAiApiRequest(data.value);
            }
        });
        if (this.message !== null) {
            this.sendMessageToWebView(this.message);
            this.message = null;
        }
    }
    async ensureApiKey() {
        this.apiKey = await this.context.globalState.get('chatgpt-api-key');
        if (!this.apiKey) {
            const apiKeyInput = await vscode.window.showInputBox({
                prompt: "请输入你的API Key",
                ignoreFocusOut: true,
            });
            this.apiKey = apiKeyInput;
            this.context.globalState.update('chatgpt-api-key', this.apiKey);
        }
    }
    async sendOpenAiApiRequest(prompt, code) {
        await this.ensureApiKey();
        // Create question by adding prompt prefix to code, if provided
        const question = (code) ? `${prompt}: ${code}` : prompt;
        if (!this.webView) {
            await vscode.commands.executeCommand('chatgpt-vscode-plugin.view.focus');
        }
        else {
            this.webView?.show?.(true);
        }
        let response = '';
        this.sendMessageToWebView({ type: 'addQuestion', value: prompt, code });
        try {
            let currentMessageNumber = this.message;
            let completion = await (0, undici_1.fetch)('https://api.openai.com/v1/chat/completions', {
                method: 'POST',
                body: JSON.stringify({
                    model: "text-davinci-003",
                    messages: [{ role: "user", content: question }],
                    temperature: 0.7
                }),
                headers: {
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    "Content-Type": 'application/json',
                    Authorization: 'Bearer ' + this.apiKey,
                },
            });
            if (this.message !== currentMessageNumber) {
                return;
            }
            if (completion.status !== 200) {
                vscode.window.showErrorMessage("GPT:请求失败，errCode=" + completion.status + "");
                return;
            }
            response = completion?.data.choices[0]?.text || '';
            const REGEX_CODEBLOCK = new RegExp('\`\`\`', 'g');
            const matches = response.match(REGEX_CODEBLOCK);
            const count = matches ? matches.length : 0;
            if (count % 2 !== 0) {
                response += '\n\`\`\`';
            }
            this.sendMessageToWebView({ type: 'addResponse', value: response });
        }
        catch (error) {
            vscode.window.showErrorMessage("GPT:请求失败.errMessage=", JSON.stringify(error.message));
            this.sendMessageToWebView({ type: 'addResponse', value: '你好！感谢您与我展开友好对话。我将尽力遵守您的要求，避免涉及法律、政治以及与中国官员有关的话题。有什么我可以帮助您的呢？无论是问题、讨论还是闲聊，我都愿意参与。' });
            return;
        }
    }
    sendMessageToWebView(message) {
        if (this.webView) {
            this.webView?.webview.postMessage(message);
        }
        else {
            this.message = message;
        }
    }
    getHtml(webview) {
        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'resource', 'webview.js'));
        const stylesMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'resource', 'webview.css'));
        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<link href="${stylesMainUri}" rel="stylesheet">
				<script src="https://cdn.jsdelivr.net/npm/marked/marked.min.js"></script>
				<script src="https://cdn.tailwindcss.com"></script>
			</head>
			<body class="overflow-hidden">
				<div class="flex flex-col h-screen">
					<div class="flex-1 overflow-y-auto" id="qa-list"></div>
					<div id="in-progress" class="p-4 flex items-center hidden">
					</div>
					<div class="p-4 flex items-center">
						<div class="flex-1">
							<textarea
								type="text"
								rows="2"
								class="border p-2 w-full"
								id="question-input"
								placeholder="Ask a question..."
							></textarea>
						</div>
						<button style="background: var(--vscode-button-background)" id="ask-button" class="p-2 ml-5">发送</button>
						<button style="background: var(--vscode-button-background)" id="clear-button" class="p-2 ml-3">清空</button>
					</div>
				</div>
				<script src="${scriptUri}"></script>
			</body>
			</html>`;
    }
}
exports.default = ChatGptViewProvider;
//# sourceMappingURL=view_provider.js.map

src/view_provider.ts

import * as vscode from 'vscode';
import { fetch } from 'undici';

export default class ChatGptViewProvider implements vscode.WebviewViewProvider {
    private webView?: vscode.WebviewView;
    private apiKey?: string;
    private message?: any;

    constructor(private context: vscode.ExtensionContext) { }

    public resolveWebviewView(
        webviewView: vscode.WebviewView,
        _context: vscode.WebviewViewResolveContext,
        _token: vscode.CancellationToken,
    ) {
        this.webView = webviewView;
        webviewView.webview.options = {
            enableScripts: true,
            localResourceRoots: [this.context.extensionUri]
        };

        webviewView.webview.html = this.getHtml(webviewView.webview);

        webviewView.webview.onDidReceiveMessage(data => {
            if (data.type === 'askChatGPT') {
                this.sendOpenAiApiRequest(data.value);
            }
        });

        if (this.message !== null) {
            this.sendMessageToWebView(this.message);
            this.message = null;
        }
    }

    public async ensureApiKey() {
        
        this.apiKey = await this.context.globalState.get('gpt-api-key') as string;
        const apiKeyInput = await vscode.window.showInputBox({
            prompt: this.apiKey?`您当前的API Key是：${this.apiKey},您可以重新输入并进行更新`:"请输入你的API Key",
            ignoreFocusOut: true,
        });
        if (apiKeyInput?.length) {
            this.apiKey = apiKeyInput!;
            this.context.globalState.update('gpt-api-key', this.apiKey);
        }
    }

    public async sendOpenAiApiRequest(prompt: string, code?: string) {
        this.apiKey = await this.context.globalState.get('gpt-api-key') as string;
        if(!this.apiKey) {
            await this.ensureApiKey();
        }

        const question = (code) ? `${prompt}: ${code}` : prompt;

        if (!this.webView) {
            await vscode.commands.executeCommand('chatgpt-vscode-plugin.view.focus');
        } else {
            this.webView?.show?.(true);
        }

        let response: String = '';

        this.sendMessageToWebView({ type: 'addQuestion', value: prompt, code });
         try {
            let currentMessageNumber = this.message;
            let completion =  await fetch('https://api.openai.com/v1/chat/completions', {
                    method: 'POST',
                    body: JSON.stringify({
                        model: "text-davinci-003",
                        messages: [{ role: "user", content: question }],
                        temperature: 0.7
                    }),
                    headers: {
                        // eslint-disable-next-line @typescript-eslint/naming-convention
                        "Content-Type": 'application/json',
                        Authorization: 'Bearer ' + this.apiKey,
                    },
                }) as any;

            if (this.message !== currentMessageNumber) {
                return;
            }

            if(completion.status !== 200) {
                vscode.window.showErrorMessage("GPT:请求失败，errCode=" + completion.status + "");
                return;
            }
            
            response = completion?.data.choices[0]?.text || '';
            const REGEX_CODEBLOCK = new RegExp('\`\`\`', 'g');
            const matches = response.match(REGEX_CODEBLOCK);
            const count = matches ? matches.length : 0;
            if (count % 2 !== 0) {
                response += '\n\`\`\`';
            }
            this.sendMessageToWebView({ type: 'addResponse', value: response });
        } catch (error: any) {
            vscode.window.showErrorMessage("GPT:请求失败.errMessage=", JSON.stringify(error.message));
            this.sendMessageToWebView({ type: 'addResponse', value: '你好！感谢您与我展开友好对话。我将尽力遵守您的要求，避免涉及法律、政治以及与中国官员有关的话题。有什么我可以帮助您的呢？无论是问题、讨论还是闲聊，我都愿意参与。' });
            return;
        }
    }

    public sendMessageToWebView(message: any) {
        if (this.webView) {
            this.webView?.webview.postMessage(message);
        } else {
            this.message = message;
        }
    }

    private getHtml(webview: vscode.Webview) {

        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'resource', 'webview.js'));
        const stylesMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'resource', 'webview.css'));

        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<link href="${stylesMainUri}" rel="stylesheet">
				<script src="https://cdn.jsdelivr.net/npm/marked/marked.min.js"></script>
				<script src="https://cdn.tailwindcss.com"></script>
			</head>
			<body class="overflow-hidden">
				<div class="flex flex-col h-screen">
					<div class="flex-1 overflow-y-auto" id="qa-list"></div>
					<div id="in-progress" class="p-4 flex items-center hidden">
					</div>
					<div class="p-4 flex items-center">
						<div class="flex-1">
							<textarea
								type="text"
								rows="2"
								class="border p-2 w-full"
								id="question-input"
								placeholder="Ask a question..."
							></textarea>
						</div>
						<button style="background: var(--vscode-button-background)" id="ask-button" class="p-2 ml-5">发送</button>
						<button style="background: var(--vscode-button-background)" id="clear-button" class="p-2 ml-3">清空</button>
					</div>
				</div>
				<script src="${scriptUri}"></script>
			</body>
			</html>`;
    }
}